File: ZemljaSlova.Model/Discount.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace ZemljaSlova.Model
{
    public class Discount
    {
        public int Id { get; set; }

        public decimal DiscountPercentage { get; set; }

        public DateTime StartDate { get; set; }

        public DateTime EndDate { get; set; }

        public string? Code { get; set; }

        public string? Name { get; set; }

        public string? Description { get; set; }

        public DiscountScope Scope { get; set; }

        public int UsageCount { get; set; }

        public int? MaxUsage { get; set; }

        public bool IsActive { get; set; }

        public virtual ICollection<OrderItem> OrderItems { get; set; } = new List<OrderItem>();

        public virtual ICollection<Order> Orders { get; set; } = new List<Order>();
    }

    public enum DiscountScope
    {
        Book = 1,
        Order = 2
    }
}

File: UI/zemljaslova_mobile/lib/models/discount.dart
import 'package:json_annotation/json_annotation.dart';

part 'discount.g.dart';

enum DiscountScope {
  @JsonValue(1)
  book,
  @JsonValue(2)
  order,
}

@JsonSerializable()
class Discount {
  final int id;
  final double discountPercentage;
  final DateTime startDate;
  final DateTime endDate;
  final String? code;
  final String? name;
  final String? description;
  final DiscountScope scope;
  final int usageCount;
  final int? maxUsage;
  final bool isActive;

  Discount({
    required this.id,
    required this.discountPercentage,
    required this.startDate,
    required this.endDate,
    this.code,
    this.name,
    this.description,
    required this.scope,
    required this.usageCount,
    this.maxUsage,
    required this.isActive,
  });

  /// Returns true if the discount is currently valid
  bool get isValid {
    final now = DateTime.now();
    return isActive && 
           startDate.isBefore(now) && 
           endDate.isAfter(now) &&
           (maxUsage == null || usageCount < maxUsage!);
  }

  factory Discount.fromJson(Map<String, dynamic> json) => _$DiscountFromJson(json);

  Map<String, dynamic> toJson() => _$DiscountToJson(this);
}
